### PR TITLE
Implement a custom keyboard, closes #29 and #251

### LINKED ISSUE #29: Custom keyboard login
- Parent issues: -
- Relevant issues: -
- Estimated effort: -

* * *

## User story
As a user I want to use a custom keyboard so that Google's or custom keyboards cannot read my passwords

## What
This issue demands a custom keyboard to type in text

## Why
This issue is needed for the project because third-party keyboard can monitor what the user is typing

## How
This feature can be added by implementing a custom keyboard for this application

## Definition of done
This task can be considered done when

- The custom keyboard is opened whenever something is typed
- The keyboard has no noticeable lag
- The user is able to type all standard characters by using the custom keyboard
- The custom keyboard can be closed when the user is ready typing

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: #29 #251 
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository a custom keyboard with limited keys and disables copying/pasting from LoginActivity

## Why
This Pull Request is needed because keyboards like Swiftkey can track what the user is typing and therefore we need (like all secure apps like those from banks) a custom keyboard that won't track what the user is typing. Also the keyboard has a limited set of keys so that the user cannot type semicolons and spaces, as required by #251 . The copying/pasting is disabled to ensure that there's no way a space or semicolon can be entered.

## How
This feature can be viewed/tested within the project by testing typing the password in LoginActivity thoroughly and verifying that you cannot copy/paste the password from the EditText.

## Alternative implementation
Other implementations that I've have considered are

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 1]
```diff
@@ -42,18 +53,57 @@ protected void onCreate(final Bundle savedInstanceState) {
...
+        });
+        editPassword.setOnClickListener(this::showCustomKeyboard);
+        editPassword.setCustomSelectionActionModeCallback(new LoginActivity.EditPasswordSelectionCallback());
+
+        this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
+
         LOGGER.info("Activity created");
     }
 
+    public void hideCustomKeyboard() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Missing JavaDoc on multiple occasions in this file 😉

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 2]
```diff
@@ -42,18 +53,57 @@ protected void onCreate(final Bundle savedInstanceState) {
...
+            if (hasFocus) {
+                this.showCustomKeyboard(v);
+            } else {
+                this.hideCustomKeyboard();
+            }
+        });
+        editPassword.setOnClickListener(this::showCustomKeyboard);
+        editPassword.setCustomSelectionActionModeCallback(new LoginActivity.EditPasswordSelectionCallback());
+
+        this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Can you add a comment on what this does? 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 3]
```diff
@@ -42,18 +53,57 @@ protected void onCreate(final Bundle savedInstanceState) {
...
+
+        editPassword.setOnFocusChangeListener((v, hasFocus) -> {
+            if (hasFocus) {
+                this.showCustomKeyboard(v);
+            } else {
+                this.hideCustomKeyboard();
+            }
+        });
+        editPassword.setOnClickListener(this::showCustomKeyboard);
+        editPassword.setCustomSelectionActionModeCallback(new LoginActivity.EditPasswordSelectionCallback());
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: What do we need this for if the `EditPasswordSelectionCallback` class returns false for all the methods it implements?
 Could you elaborate a bit on what the methods in this class do?

File: app/src/main/res/layout/preview.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:gravity="center"
+    android:background="#ffff00"
+    android:textStyle="bold"
+    android:textSize="30sp"
+    >
```
Comment (ericcornelissen) [inline app/src/main/res/layout/preview.xml]: Could you move this `>` to the previous line 😇

File: app/src/main/res/drawable/key_background.xml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:drawable="@drawable/button_pressed" android:state_pressed="true" />
+    <item android:drawable="@drawable/button_pressed" android:state_focused="true" />
+    <item android:drawable="@drawable/button_pressed" android:state_hovered="true" />
```
Comment (ericcornelissen) [inline app/src/main/res/drawable/key_background.xml]: Not sure if you want the same clover on hover 🤔

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Added a comment about that

File: app/src/main/res/drawable/key_background.xml (hunk 5 above)
Comment (jverbraeken) [inline app/src/main/res/drawable/key_background.xml]: I think that the change is very, very small that people are using a cursor on their smartphone

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java (hunk 3 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: 👍

File: app/src/main/res/drawable/key_background.xml (hunk 5 above)
Comment (ericcornelissen) [inline app/src/main/res/drawable/key_background.xml]: I think so too, but if you took the time to add it, I don't think it is that much effort to make a slightly brighter color for when it is hovered

**Comment by clenrock** [comment]
image
I don't think that this is nice enough. It also needs a back button.
But in general I would say that the normal keyboard is nicer and that we should keep that

**Comment by ericcornelissen** [comment]
I'm not able to run the app on this branch, not sure why though 😕 See:
capture

File: app/src/main/java/com/nervousfish/nervousfish/activities/CustomKeyboardHelper.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }
+    }
+
+    private void clickListener(final View view) {
+        this.keyboardView.setVisibility(View.VISIBLE);
+        this.keyboardView.setEnabled(true);
+        this.hideRealKeyboard(view);
+    }
+
+    private void hideRealKeyboard(final View view) {
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CustomKeyboardHelper.java]: Why is this method needed when you already hide it in the constructor?

File: app/src/main/res/drawable/button_pressed.xml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layer-list xmlns:android="http://schemas.android.com/apk/res/android">
+    <item>
+        <shape android:shape="rectangle">
+            <solid android:color="@color/color_primary_dark" />
+            <corners
+                android:bottomLeftRadius="4dp"
+                android:bottomRightRadius="4dp"
+                android:topLeftRadius="4dp"
+                android:topRightRadius="4dp" />
```
Comment (TheBonheurs) [inline app/src/main/res/drawable/button_pressed.xml]: shouldn't you add these to dimens?

File: app/src/main/java/com/nervousfish/nervousfish/activities/CustomKeyboardHelper.java (hunk 6 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CustomKeyboardHelper.java]: Oops 😅 put it there while trying to get the custom keyboard working in the `CreateProfileActivity` but forgot to remove it

File: app/src/main/res/drawable/button_pressed.xml (hunk 7 above)
Comment (ericcornelissen) [inline app/src/main/res/drawable/button_pressed.xml]: I think so yes 🤔 Nice find!

**Pull request merged**